codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerOperationHandlerGenerator.kt

```diff
@@ -132,13 +135,19 @@ class ServerOperationHandlerGenerator(
         } else {
             symbolProvider.toSymbol(operation.outputShape(model)).fullName
         }
+        val streamingBodyTraitBounds = if (operation.inputShape(model).hasStreamingMember(model)) {
+            "B: Into<#{SmithyHttp}::byte_stream::ByteStream>,"
```

Comment:
The gut feeling is that we should re-export ByteStream from smithy-http into smithy-http-server. This way we can "control" the API a little more and prevent breaking changes since the server codegen does not own smithy-http.

Reply:
@david-perez thoughts?

Reply:
we re-export ByteStream from generated services FYI

Reply:
that's right - I see it re-exported in the lib.rs of any generated server crate like so:
```
/// Crate version number.
pub static PKG_VERSION: &str = env!("CARGO_PKG_VERSION");
pub use aws_smithy_http::byte_stream::ByteStream;
pub use aws_smithy_http::result::SdkError;
pub use aws_smithy_types::Blob;
pub use aws_smithy_types::DateTime;
```

Reply:
@david-perez notice that also SdkBody is exported there ^ not sure if there's a reason behind it or not

Reply:
For now I re-exported from aws_smithy_http_server too and used the re-exported type instead.

Reply:
@guymguym So @crisidev and I took the design decision a while back to **not** use `SdkBody` in the server, since it's a special body type that caters to the [AWS Rust SDK's](https://github.com/awslabs/aws-sdk-rust) needs for _dispatching_ HTTP requests. From a technical perspective there's no impediment to using it in the server at the moment, but in the future this type might evolve further to satisfy the requirements of the AWS Rust SDK.

We also considered having a server body type of our own, but eventually we decided that since we don't have any special requirements, we should write our interfaces against the trait [`http_body::Body`](https://docs.rs/http-body/latest/http_body/trait.Body.html) wherever possible. We also decided to reexport an HTTP body trait from `aws_smithy_http_server` that endorsed [`hyper::body::HttpBody`](https://docs.rs/hyper/latest/hyper/body/trait.HttpBody.html), and unboxed and boxed dyn body types that endorsed [`hyper::body::Body`](https://docs.rs/hyper/latest/hyper/body/struct.Body.html) and [`http_body::comibnators::UnsyncBoxBody`](https://docs.rs/http-body/0.4.4/http_body/combinators/struct.UnsyncBoxBody.html), respectively. The rationale was that these crates have good stable types that are used throughout the Rust ecosystem, and as such users would have a better experience if they wanted to use these bodies in other libraries, and we retain some degree of autonomy by reexporting them in case in the future we'd like to expose a body type of our own. It's also the approach that other frameworks took, [like `axum`](https://docs.rs/axum/latest/axum/body/index.html).

Truth be told, Smithy service implementers shouldn't have to worry too much about body types, since they are hidden well inside the framework and they just code against modeled input and output types, but it's inevitable to expose them in the public API (e.g. in the Router). So in that sense we differ from the typical web framework. Another place where they seep in is here: in streaming requests and responses. I didn't even realize we were exposing `SdkBody` via `ByteStream` in service's `lib.rs`; I guess because we hadn't implemented streaming so we've never used `ByteStream`.

But now that you're adding streaming support, we need to decide and commit more firmly on this design aspect. Personally, I think the server should _still continue_ to use the `http_body` and `hyper` types and not expose `SdkBody`, at least until we have a need for `SdkBody` or our own server body type. This means we can't expose `ByteStream`, since it's tied to `SdkBody`, but the `ByteStream` abstraction is perfect and just what we need for this feature (aside: this is a recurring pattern you'll notice in the codebase, since it got started for the needs of the AWS Rust SDK and we've been slowly refactoring it and stealing useful utilities for the server side where applicable). So to move forward and land this feature, I think I lean towards wrapping `ByteStream` in `aws_smithy_http_server` to hide away the fact that it's using `SdkBody` underneath, and convert between body types in our implementation if we need to until we figure out how best to write it for service implementors. For instance, I haven't looked too closely at `ByteStream` implementation, but one thing I'm missing that I foresee service implementors will need is the ability for them to create and return a `ByteStream` from an arbitrary [`std::stream::Stream`](https://doc.rust-lang.org/std/stream/trait.Stream.html) from their handlers. We can also have the server's `ByteStream` start very simple (just taking in `&'static [u8]`, for instance) and slowly expand it in future PRs. This is where **you can greatly help us**, since we don't have any customers that are using streaming in their Rust services. Whatever your application's needs will be in this regard, we can slowly add them to the streaming support.

But what I'm certain of is that we should not expose / endorse **both** `SdkBody` and the other body types, since it will make things confusing for users.

---

Sorry for the long comment and thanks for the PR! You're helping us tremendously with these and I'm excited to see what you will build with this project.

Reply:
@david-perez Thanks for all the context! That's very helpful for me as an external collaborator. 

I understand you design direction, but this also requires changing the type in the generated structs. Do you want me to take a stab at creating the server byte_stream in this PR? Or do you prefer to commit and that I make another PR for that change?

Reply:
@guymguym 

> but this also requires changing the type in the generated structs

That's right. In struct generation, instead of generating:

```rust
#[allow(missing_docs)] // documentation missing in model
#[non_exhaustive]
pub struct StreamingTraitsOutput {
    #[allow(missing_docs)] // documentation missing in model
    pub foo: std::option::Option<std::string::String>,
    #[allow(missing_docs)] // documentation missing in model
    pub blob: aws_smithy_http::byte_stream::ByteStream,
}
```

we would generate `pub blob: aws_smithy_http_server::byte_stream::ByteStream`.

> Do you want me to take a stab at creating the server byte_stream in this PR?

Yeah, let's try that first. Make separate commits so that in case the diff gets too big we can easily split the PR and merge what you've done already.

Reply:
Alrighty then. Lets go for it! I'll update soon and ping for your review.

Reply:
Conclusion in https://github.com/awslabs/smithy-rs/pull/1023#discussion_r784751006. Resolving.